### PR TITLE
KTOR-7820 Dependency injection design

### PR DESCRIPTION
**Pull request by bjhham**
This is a proposal to resolve:

https://youtrack.jetbrains.com/issue/KTOR-6621/Make-Dependency-Injection-Usage-Simple

The goal here is to provide an easy way to introduce dependency injection for new users, especially those who are accustom to the ease of frameworks like Spring.  I tried to address any drawbacks found in other solutions and keep to the Ktor philosophy as much as possible.

### CODE REVIEW

**Comment by osipxd** [comment]
[Rendered preview](https://github.com/ktorio/ktor-klip/blob/bjhham/dependency-injection/proposals/0001-dependency-injection.md)

File: proposals/0001-dependency-injection.md [hunk 1]
```diff
@@ -0,0 +1,527 @@
+|             |                                                                                                                                       |
+|-------------|---------------------------------------------------------------------------------------------------------------------------------------|
+| Feature     | Dependency Injection                                                                                                                  |
+| Submitted   | 2024-12-06                                                                                                                            |
+| Accepted    | No                                                                                                                                    |
+| Issue       | https://youtrack.jetbrains.com/issue/KTOR-6621/Make-Dependency-Injection-Usage-Simple                                                 |
```
Comment (osipxd) [inline proposals/0001-dependency-injection.md]:
```suggestion
| Issue       | [KTOR-6621](https://youtrack.jetbrains.com/issue/KTOR-6621/Make-Dependency-Injection-Usage-Simple)                                    |
```

File: proposals/0001-dependency-injection.md [hunk 2]
```diff
@@ -0,0 +1,527 @@
...
+10. [Future Directions](#future-directions)
+
+<hr />
+
+# Summary
+[summary]: #summary
+
+We describe a general solution for handling [dependency injection (DI)](https://en.wikipedia.org/wiki/Dependency_injection)
+in Ktor server applications.  Our intent is to provide a simple, performant interface, while avoiding the drawbacks that
+that can arise when working with the automatic configuration of dependencies.
```
Comment (osipxd) [inline proposals/0001-dependency-injection.md]:
```suggestion
in Ktor server applications.  Our intent is to provide a simple, performant interface, while avoiding the drawbacks
that can arise when working with the automatic configuration of dependencies.
```

File: proposals/0001-dependency-injection.md [hunk 3]
```diff
@@ -0,0 +1,527 @@
...
+## Resolution
+[resolution]: #resolution
+
+In practice, we can expect the declaring modules to be few, and the resolving modules to be many.
+
+Here is an example of how to resolve a dependency from the `Application` scope:
+
+```kotlin
+fun Application.configure() {
+    val repository by resolve<Repository<User>>()
```
Comment (osipxd) [inline proposals/0001-dependency-injection.md]: Possible alternative syntax:
```kotlin
fun Application.configure() {
    val repository by dependencies.resolve<Repository<User>>()
}
``` 
if we want to have a common entry point for both declaration and resolution APIs.

File: proposals/0001-dependency-injection.md [hunk 4]
```diff
@@ -0,0 +1,527 @@
...
+
+
+# Technical Details
+[technical-details]: #technical-details
+
+In practice, we can expect developers to include several dependency declaration modules, followed by many more 
+resolutions in the feature modules.  To handle this, the implementation will need to build a tree of dependencies 
+from the set of declaration blocks, then perform the resolutions in a blocking scope at the site of the first 
+"application-scoped" `resolve()` call.  This introduces temporal coupling between the two phases of declaration and 
+resolution, though the declarations can happen in any order.
```
Comment (osipxd) [inline proposals/0001-dependency-injection.md:348]: > This introduces temporal coupling between the two phases of declaration and 
resolution

So we can avoid this only by making `resolve` suspending, right? Could you add some context why we decided not to do this?

File: proposals/0001-dependency-injection.md [hunk 5]
```diff
@@ -0,0 +1,527 @@
...
+constructor, we can include some logging when the framework decides automatically, indicating that an explicit 
+declaration is required for overriding the behavior.  As for the choice process in the framework itself, one option
+might be to attempt the constructor with the fewest arguments.  Some further research on the topic will be necessary.
+
+## Validation
+[validation]: #validation
+
+Normally, validation occurs when calling `resolve()` in the application.  This will make an attempt to find the instance
+of the expected type in the instance repository.  If nothing was declared for this type, then it will throw an exception
+`DependencyMissingException`, unless otherwise implemented by a custom provider.
```
Comment (osipxd) [inline proposals/0001-dependency-injection.md]: We should also handle possible circular dependencies. As circular dependencies could lead to dead lock:
```kotlin
class Foo(bar: Bar)
class Bar(foo: Foo)

fun Application.configure() {
    dependencies {
        provide<Foo> { Foo(resolve<Bar>()) } // Foo waits for Bar
        provide<Bar> { Bar(resolve<Foo>()) } // Bar waits for Foo
    }
}
```

File: proposals/0001-dependency-injection.md [hunk 6]
```diff
@@ -0,0 +1,527 @@
...
+
+The API looks like this:
+
+```kotlin
+// installation
+fun Application.installDependencies() {
+    koin {
+        modules(module {
+            single { Service() }
+            factory(named("key")) { Repository(get()) }
```
Comment (osipxd) [inline proposals/0001-dependency-injection.md:91]: In Koin you can also [pass a constructor](https://github.com/InsertKoinIO/koin/blob/fa85a195e748ef14c5a3c5abe474f2bffebeb579/projects/core/koin-core/src/commonMain/kotlin/org/koin/core/module/dsl/FactoryOf.kt#L41-L44) instead of lambda. All parameters will be retrieved using `get()`:
```
factoryOf(::Repository)
```
This solution is a compromise between manual and reflective instantiation.

File: proposals/0001-dependency-injection.md (hunk 5 above)
Comment (bjhham) [reply proposals/0001-dependency-injection.md]: Oh yes, I meant to include a note on that, thanks!

File: proposals/0001-dependency-injection.md (hunk 6 above)
Comment (bjhham) [reply proposals/0001-dependency-injection.md:91]: Oh, I like that!

File: proposals/0001-dependency-injection.md (hunk 4 above)
Comment (bjhham) [reply proposals/0001-dependency-injection.md:348]: My thoughts are that during declaration they'll suspend like this, but at a certain point we have to decide that declarations are done so that we know when to fail.

File: proposals/0001-dependency-injection.md (hunk 3 above)
Comment (bjhham) [reply proposals/0001-dependency-injection.md]: I like the symmetry, though it's a tad verbose.

Comment (osipxd) [reply proposals/0001-dependency-injection.md]: Another alternative:
```kotlin
val repository: Repository<User> by dependencies
```
But it requires tweaking to support named instances.

File: proposals/0001-dependency-injection.md (hunk 6 above)
Comment (arnaudgiuliani) [reply proposals/0001-dependency-injection.md:91]: If I can add some clarification here: no reflection, it's a (dirty) trick to match the signature and fill in the needed dependencies.

Comment (osipxd) [reply proposals/0001-dependency-injection.md:91]: > If I can add some clarification here: no reflection, it's a (dirty) trick to match the signature and fill in the needed dependencies.

Thank you for the clarification! Yes, I called it "a compromise" because it is as fast as manual instantiation but lacks some features of reflective instantiation. Though, I can name only one missing feature - named instances aren’t supported.

Comment (arnaudgiuliani) [reply proposals/0001-dependency-injection.md:91]: yes, this is a point where I would need a compiler plugin or something here

**Comment by JakeWharton** [comment]
Hello!

The "current solution" examples provided for Koin and Kodein, and the proposed Ktor APIs are not dependency injection. They are service location. A single service locator instance is supplied from which dependencies can be imperatively pulled.

Let's look at the Wikipedia article you start by citing (https://en.wikipedia.org/wiki/Dependency_injection):

> dependency injection is a programming technique in which an object or function receives other objects or functions that it requires, as opposed to creating them internally

and

> Instead, the receiving "client" (object or function) is provided with its dependencies by external code (an "injector"), which it is not aware of

Neither of the examples nor the proposed API receive the objects or functions they require directly. And neither example nor the Ktor API is unaware of the source from where it resolves its dependencies.

Instead, each receives a single, framework-specific instance registry from which it pulls the dependencies that it requires as implementation detail.

Let's contrast that with the service locator pattern from Wikipedia (https://en.wikipedia.org/wiki/Service_locator_pattern):

> This pattern uses a central registry known as the "service locator", which on request returns the information necessary to perform a certain task.

That description _precisely_ matches what is going on in samples and the proposed API.

There is nothing inherently wrong with service location vs. dependency injection–it's simply a different pattern. I would strongly suggest renaming this proposal and its proposed APIs to be named after the correct pattern that it is modeling.

**Comment by bjhham** [comment]
Hi @JakeWharton, this design places more focus on the manual declaration of dependencies as with a service locator, but it's intended to be used with automatic instantiation through reflection as well ([see section](https://github.com/ktorio/ktor-klip/blob/bjhham/dependency-injection/proposals/0001-dependency-injection.md#instantiation-by-reflection)).  From what I understand, this automatic injection of types is the only meaningful distinction.

I'll make some updates to expand a bit more on this concept - the intention of the design is to separate dependency declaration and resolution, so that resolving types from your modules can be handled through any means - automatic, external, or otherwise.

**Comment by JakeWharton** [comment]
I saw that. However, that section contains this sentence:

> For the initial prototype of dependency injection in Ktor, we can forgo this automatic instantiation, but it would be useful to discuss the approach here for future reference.

Additionally, it's extremely hand-wavy on behavior, and is not a concrete part of this proposal.

This proposal is for a service locator with an imperative API. Perhaps a future proposal can layer dependency injection on top of that service locator. Almost all of the popular dependency injectors are built on top of a service locator, so this is with precedent.

By updating this proposal to accurately reflect what is being designed today we can properly set expectations and use the correct definitions for the provided behavior and API. The reflection-based can then be updated to refer to itself as dependency injection with the Wikipedia link and so on.

**Comment by bjhham** [comment]
Fair point.  I'll update the document with some greater clarification in this regard.

File: proposals/0001-dependency-injection.md (hunk 6 above)
Comment (bjhham) [reply proposals/0001-dependency-injection.md:91]: Yeah, I don't think it's possible to get around the use of annotations + reflection for named instances when injecting constructors.

Comment (arnaudgiuliani) [reply proposals/0001-dependency-injection.md:91]: Yeah, this is a question that I try to tackle for Koin too. Or perhaps a special type marker? But then your constructor property is linked to something of the framework 🤔

Comment (arnaudgiuliani) [reply proposals/0001-dependency-injection.md:91]: In spring framework, previous to annotations, we had to specify it manually

File: proposals/0001-dependency-injection.md (hunk 3 above)
Comment (bjhham) [reply proposals/0001-dependency-injection.md]: I'm thinking now...

```kotlin
// basic use case
val repository: Repository<User> by dependencies

// with named instance
val repository: Repository<User> by dependencies.named("users")

// or without property delegate
val repository: Repository<User> = dependencies.resolve()
val repository: Repository<User> = dependencies.resolve("users")
```

Comment (arnaudgiuliani) [reply proposals/0001-dependency-injection.md]: With such an approach on Koin, we propose also to use type qualifier, to use class or enum instead of strings

File: proposals/0001-dependency-injection.md [hunk 7]
```diff
@@ -0,0 +1,574 @@
...
+injection, because it would create too much incongruity with the framework.  To continue with our commitment of 
+"no magic" programmatic configuration, we'll stick with frameworks that provide programmatic means for injection.
+
+# Design Overview
+[design-overview]: #design-overview
+
+Drawing from our review of other solutions, we have compiled the following design goals and requirements for the 
+integrated dependency injection system.
+
+### Core Requirements
```
Comment (e5l) [inline proposals/0001-dependency-injection.md:161]: What would be nice to have is an API for integrations with other DIs (so the repository can be reused).

File: proposals/0001-dependency-injection.md [hunk 8]
```diff
@@ -0,0 +1,574 @@
...
+In this example, both modules are called from the main `module()`, but they can be treated as interchangeable units 
+using configuration.
+
+For example, when using YAML configuration: 
+
+```yaml
+# application.yaml / ktor / application 
+modules:
+   - io.ktor.example.ImplementationKt.declareDependencies
+   - io.ktor.example.ApplicationKt.configureApplication
```
Comment (e5l) [inline proposals/0001-dependency-injection.md:217]: Should we try registring injectables (like factories) in the configuration file?

File: proposals/0001-dependency-injection.md [hunk 9]
```diff
@@ -0,0 +1,574 @@
...
+
+## Resolution
+[resolution]: #resolution
+
+In practice, we can expect the declaring modules to be few, and the resolving modules to be many.
+
+Here are the examples of how to resolve dependencies from a Ktor module:
+
+```kotlin
+fun Application.configureApplication() {
```
Comment (e5l) [inline proposals/0001-dependency-injection.md]: Could you tell me what you think about supporting parameters injection for the modules declared in the config file?

```config.yml
dependencies:
    mongo: DatabaseKt.mongoProduction

modules:
   - ModuleKt.module
   
```

```Module.kt
fun Application.module(mongo: DataSource) {
    // ...
}
```

Comment (e5l) [reply proposals/0001-dependency-injection.md]: We also can allow registries from DI framework to contribute to shared Ktor registry, so they can benefit from the syntax we have

**Review by e5l — CHANGES_REQUESTED** [review]

**Comment by e5l** [comment]
Hey @bjhham, thanks for the proposal. I've left some comments, please take a look.

btw I think @JakeWharton made a good point about naming. Let's adjust names and mention the difference in the description explicitly!

File: proposals/0001-dependency-injection.md (hunk 7 above)
Comment (bjhham) [reply proposals/0001-dependency-injection.md:161]: This should be possible using the extension points [mentioned here]( https://github.com/ktorio/ktor-klip/blob/bjhham/dependency-injection/proposals/0001-dependency-injection.md#extensibility) though I could include a better example with Koin there.

File: proposals/0001-dependency-injection.md (hunk 8 above)
Comment (bjhham) [reply proposals/0001-dependency-injection.md:217]: Yeah, this makes sense... Though it could be a little confusing when you start mixing both methods.  I'll do some thinking on it.

File: proposals/0001-dependency-injection.md (hunk 9 above)
Comment (bjhham) [reply proposals/0001-dependency-injection.md]: Yeah, originally I was thinking of doing it this way.  I think it makes sense to offer this as an alternative.  One possible issue with this is that the code for calling into the modules in the core, but we're aiming to make this an extension / plugin.

Comment (e5l) [reply proposals/0001-dependency-injection.md]: Yep, we can make the Environment implementation pluggable

**Comment by bjhham** [comment]
> btw I think @JakeWharton made a good point about naming. Let's adjust names and mention the difference in the description explicitly!

I've since updated the design to provide some proper dependency injection tooling out of the box, so I think the naming should be correct now.  If there's any naming that looks off, could you specify the items that need review?

File: proposals/0001-dependency-injection.md [hunk 10]
```diff
@@ -0,0 +1,602 @@
...
+   override fun <T : Any> get(key: DependencyKey): T =
+      koin.getOrNull(key.type.type, key.name?.let(::StringQualifier))
+         ?: base.get(key)
+}
+```
+
+This allows resolving dependencies that are provided by Koin modules.
+The full example can be found at [ktor-chat/tree/dependency-injection-koin-ext](https://github.com/bjhham/ktor-chat/tree/dependency-injection-koin-ext)
+
+## Testing
```
Comment (IRus) [inline proposals/0001-dependency-injection.md:578]: This KLIP does not explain why a Ktor application needs DI. The Motivation section simply states that users are asking for it, so it should be added, and this KLIP proposes adding it. Unfortunately, it lacks an explanation of what DI makes easier or better.

For me, DI is a way to eliminate tightly coupled code and simplify the testing process. Here, I would like to see more examples and evidence that this approach will indeed work well for testing.

The current test example is overly simplistic—so simple that it seems like there’s no real advantage to using DI. Manual DI, in this case, could be lighter, simpler, and offer compile-time guarantees:

```
fun Application.module() {
    val dataSource = HikariDataSource()
    val users = Repository(dataSource)

    configureMyEndpoint(users)
}

fun Application.configureMyEndpoint(
    users: Repository
) {
    get("/users") {
        call.respond(users.list())
    }
}

fun main() {
    embeddedServer(Netty, port = 8080, host = "0.0.0.0", module = Application::module)
        .start(wait = true)
}

@Test
fun verifyEndpoint() = testApplication {
    val users = kmock<Repository>().also { it._fetch returns listOf(myEntity) }
    configureMyEndpoint(users)

    // execute tests with client
}
```

What really needs to be demonstrated is how DI works in a modular structure where there are many modules and dependencies between them. In this context, I assume that more than one `dependencies` block is allowed in the application. In such a setup, it’s unclear how to write an integration test for a handler. Specifically, how do you enumerate all the modules with declarations that the current handler depends on? This is not described. If all modules are loaded in tests, this creates unnecessary overhead and doesn’t allow testing a module in isolation.

It’s also important to ensure that multiple tests can run in parallel, i.e., there should not be a single global context, as seen in basic Koin examples (https://ruslan.ibragimov.by/2024/08/11/koin-puzzler).

For my projects, I wrap Ktor handlers and features in components that are then injected into the server (https://github.com/Heapy/awesome-kotlin/blob/main-new/app-backend/src/main/kotlin/server/KtorServerModule.kt#L39). This way, Ktor doesn’t become the foundation of the application, like some other frameworks, but rather a component of the application that handles a part of the HTTP server/client functionality. This provides flexibility, such as creating a server and passing in only a single route that’s important to test in a given scenario.

Comment (bjhham) [reply proposals/0001-dependency-injection.md:578]: Hey, thanks for the feedback.  I can add a few sentences to describe the value of DI in relation to Ktor projects - I had omitted for brevity.

Could you elaborate on what is missing for you in the example project?  There are a few tests that show how to populate your application with some modules omitted and others mocked while having interrelated dependencies across multiple gradle modules, here is an example ([MessagesTest](https://github.com/bjhham/ktor-chat/blob/287f48a367f3f4be68ab3f7cee0cbabaef10479c/server/rest/test/io/ktor/chat/MessagesTest.kt)) where authentication and the repository implementations are replaced for testing.  The tests can be run in parallel as well - the dependencies are scoped to the Ktor server instance and not the process.

Comment (IRus) [reply proposals/0001-dependency-injection.md:578]: @bjhham Thanks.

> Could you elaborate on what is missing for you in the example project?

Basically I don't think that defining all required modules manually in test either practical or eliminates coupling https://github.com/bjhham/ktor-chat/blob/287f48a367f3f4be68ab3f7cee0cbabaef10479c/server/rest/test/io/ktor/chat/MessagesTest.kt#L19 

I using DI on top of delegated properties at work where at first I have to create all modules in tests manually, i.e:

```kotlin
val module = Module1(Module2(Module3(), Module4()))
```

It proved itself as working solution, but it's creates a lot of conflicts and code change once we need to add new module in tree. So I endup having small reflective part to build this three by just specifying Module1:

```kotlin
// create module by traversing primary constructors. It is a requirement to modules to depend only on other modules or nothing
val module: Module1 = buildModule<Module1>() 

// create basically Map<Kclass, Instance> of all modules required for Module1
val modules: Modules = buildModules<Module1>() 
val module3 = modules<Module3>()
```

Opensource version of this: https://github.com/Heapy/komok/blob/main/komok-tech-to-be-injected/src/test/kotlin/io/heapy/komok/tech/di/delegate/Example.kt#L109

So you can clearly see that adding Module5 as constructor parameter for Module4 will require to update all the tests in case of proposal and manually building module tree. So what's the point for implicit Map<KType, T>, when you can replace it with compile-time, almost free implementation on top of delegates? 

I think proposed approach making nice API for defining beans, but then it's lacks of a way to scale it for applications where DI would make sense.

I can also share another approach I'm using on top of plain classes and lazy delegates, but with layer of KSP code-generation specifically for tests for overriding modules and specific beans. It's also a good example of module management when user can add new modules as dependencies to existing modules without modifying any additional production or testing code: https://github.com/Heapy/awesome-kotlin/blob/main-new/app-backend/src/test/kotlin/usecases/robots_txt/RobotsTxtRouteTest.kt#L15

TL;DR; My concern that this:

```
Application::rootModule,
Application::restModule,
Application::mockAuth,
Application::mockMessagesRepository,
Application::messagesModule,
```
 
for DI not better than this (as example):
 
```
Module1(Module2(Module3()))
 
class Module {
  val repo by bean { 
      Repo()
  }
} 
```
 
But lacks compile-time safety at the same time.

Comment (bjhham) [reply proposals/0001-dependency-injection.md:578]: Hey, I introduced a section on parameter injection of modules - we'll have the option of declaring modules (still as functions) with parameters that will be resolved by the framework automatically when configured by reference, or you can call them as functions from code to leverage some compile-time safety.  We're also planning on introducing custom inspections for discovering missing dependencies for the gradle plugin and/or the IDE plugin so that you can get the loose coupling of runtime resolution with the safety of compile time checks.

Comment (IRus) [reply proposals/0001-dependency-injection.md:578]: @bjhham Hey, thanks for the update.

https://github.com/ktorio/ktor-klip/blob/bjhham/dependency-injection/proposals/0001-dependency-injection.md#by-parameters

```
fun Application.configureRouting(
   users: Repository<User>,
   mongo: DataSource,
   messages: Repository<Message>,
) {
   routing {
      get("/users") {
         call.respond(users.list())
      }
      // ...
   }
}
```

## Case 1

Not sure how this addressing use case when most of beans should be initialized, and just one of dependencies mocked. 

Example: 

`Repository<User>` needs `Clock` as dependency to get current time. In particular test we want to provide custom version of `Clock` which will be a test version (user can control `Clock` from test).

So tests need just standard initialized `DataSource` and `Repository<Message>`,  but want to change how one of beans injected into `Repository<User>` initialized.

## Case 2

Adding new dependency to module will change function signature and all tests needs to be updated, even if they fine with standard implementation.

File: proposals/0001-dependency-injection.md (hunk 8 above)
Comment (arnaudgiuliani) [reply proposals/0001-dependency-injection.md:217]: from there, you load a new extension to load definitions right?

File: proposals/0001-dependency-injection.md [hunk 11]
```diff
@@ -0,0 +1,633 @@
...
+This way, we can introduce a dependency inversion where the `declareDependencies` module can live on a different runtime
+classpath than the application module `configureApplication`, which has no knowledge of the implementation details.
+
+In the next two sections, we'll provide the API design for both declaring and resolving dependencies.
+
+## Declaration
+[declaration]: #declaration
+
+As with declaring extensions in a Ktor application, users can expect to declare their dependencies from the
+same scope: `Application`.
```
Comment (arnaudgiuliani) [inline proposals/0001-dependency-injection.md:228]: Don't you consider any other scope besides the main default "Application" one? No request scope definitions?

File: proposals/0001-dependency-injection.md [hunk 12]
```diff
@@ -0,0 +1,633 @@
...
+```kotlin
+fun Application.rootModule() {
+   install(Koin) {
+      modules(module {
+         single<Algorithm>(named("hash")) {
+            Algorithm.HMAC256(property("security.secret"))
+         }
+      })
+   }
+   install(DI) {
```
Comment (arnaudgiuliani) [inline proposals/0001-dependency-injection.md:560]: Could the `install(Koin)` provides also the `install(DI)` part? to avoid having to install both, and keep writing only `install(Koin)`?

File: proposals/0001-dependency-injection.md [hunk 13]
```diff
@@ -0,0 +1,633 @@
...
+Most phases in the Ktor application lifecycle include some means of registering hooks, like startup or shutdown.  Until 
+we hear a compelling use-case for invoking some code when dependencies are declared or resolved, we'll avoid introducing 
+this functionality.
+
+## Performance
+[performance]: #performance
+
+The general performance bottlenecks found in DI frameworks are caused from heavy up-front scans, heavy use of 
+reflection, and repeated calls to slow instantiation.  These issues will be avoided in this design by using Kotlin's 
+compile-time type inference (using `typeOf<T>()`), minimal use of reflection, and by only using singletons.
```
Comment (arnaudgiuliani) [inline proposals/0001-dependency-injection.md:479]: My last attempt of `typeOf<T>()` implied performance penalties. On ktor server, the performances are ok. But perhaps to check for client injection?

File: proposals/0001-dependency-injection.md [hunk 14]
```diff
@@ -0,0 +1,633 @@
...
+`resolve()` to `create()`.  This suggests that, rather than searching for the instance, we'll create it from its 
+parameters.  Anywhere resolution takes place, the create function can be called.
+
+The desired extent of automation in a dependency injection framework is a matter of personal preference.  For that 
+reason, we will include means for configuring the default behaviors.
+
+### Configuration
+[configuration]: #configuration
+
+#### Create behavior
```
Comment (arnaudgiuliani) [inline proposals/0001-dependency-injection.md:392]: for the `create` function you want to mix DSL and annotations? 

I don't understand well the difference between `provide<Repository<Message>>(MessageRepository::class)` and `provide { create(::RoomRepository) }` example

File: proposals/0001-dependency-injection.md (hunk 13 above)
Comment (e5l) [reply proposals/0001-dependency-injection.md:479]: This issue is known Kotlin issue for Android. It is reported and should be fixed.

File: proposals/0001-dependency-injection.md (hunk 10 above)
Comment (bjhham) [reply proposals/0001-dependency-injection.md:578]: Oh sorry I misunderstood the direction you were going with - I think I see what you're saying now, you want to pinpoint which dependencies you want to override but retain a standard set of modules as defined in the production application without listing them out in every test.

Something like:

```
fun test() = testApplication {
    dependencies {
         provide<Dependency> { Implementation() }
    }
    installOtherModulesFromConfig()
    
    // do test
}
```

For this we can introduce some conflict resolution so that our test application module dependency declarations take precedence over the declarations in the production modules.  The syntax here could use a little finessing, but I think it ought to work nicely for what you're describing.

File: proposals/0001-dependency-injection.md (hunk 12 above)
Comment (bjhham) [reply proposals/0001-dependency-injection.md:560]: Yes, it certainly could!

File: proposals/0001-dependency-injection.md (hunk 14 above)
Comment (bjhham) [reply proposals/0001-dependency-injection.md:392]: > for the create function you want to mix DSL and annotations?

Yes, having a few annotations seems the simplest way for resolving ambiguity in these cases.  We could also have some kind of declarative rules system if we really want to avoid annotations, but I think it's probably better to stick with what's familiar here for now.

> I don't understand well the difference between provide<Repository<Message>>(MessageRepository::class) and provide { create(::RoomRepository) } example

Sorry I was a little vague there - the class reference at the top level should be the same as calling create, the main reason for the create function is further down the stack in places like `provide { RoomRepository(create(::PostgresDataSource), resolve<SomeOtherArg>()) }`.

File: proposals/0001-dependency-injection.md (hunk 12 above)
Comment (arnaudgiuliani) [reply proposals/0001-dependency-injection.md:560]: Ok cool

File: proposals/0001-dependency-injection.md (hunk 13 above)
Comment (arnaudgiuliani) [reply proposals/0001-dependency-injection.md:479]: Ok 👍 I will test again then

File: proposals/0001-dependency-injection.md (hunk 14 above)
Comment (arnaudgiuliani) [reply proposals/0001-dependency-injection.md:392]: Yes, I guess it's to avoid too much usages of resolve<> around a definition.

File: proposals/0001-dependency-injection.md (hunk 10 above)
Comment (IRus) [reply proposals/0001-dependency-injection.md:578]: @bjhham Yes, that’s exactly where I’m heading. Integration testing is crucial, and Spring provides one of the most convenient infrastructures for it. However, their approach to contexts doesn’t make it easy to balance ease of writing and performance simultaneously. For example, when mocking something in a parallel test case, we often need to use @DirtiesContext. It’s a trade-off between isolation and performance on their side. Isolated contexts and, ideally, some form of reconciliation are important for enabling parallel test execution.

What hides scalability issues is:

```kotlin
installOtherModulesFromConfig()
```

This raises questions such as:
1. Does it install all available modules, i.e., execute all of them? If so, is the dependency tree lazy or eager? How would that affect the performance of larger applications? And if it installs all of them, are contexts cached between tests? Context setup is not cheap, especially when dealing with components like database connections that may take seconds to establish. Ideally, we’d want to share such resources across all tests.
2. If module selection inside this function is manual, we return to a previous discussion about its drawbacks.

In my experience, building some form of Dependency Injection (DI) is relatively easy. What’s difficult is designing it in a way that allows building small, isolated parts of the context while optionally reusing a “base” context for heavy beans, handling the bean lifecycle (start/stop), and keeping it compile-time safe.

Comment (bjhham) [reply proposals/0001-dependency-injection.md:578]: So for `installOtherModulesFromConfig()`, you could use the standard config file to load all the modules, or import some other test config, or simply introduce one module at a time manually.  We don't have any means for sharing context between test or production servers at the moment, so the only way to avoid the performance hit for a heavy instantiation would be to supply a static instance for the dependency.  We could probably introduce some kind of cache mechanism for testing, at least when using configuration files.

Comment (IRus) [reply proposals/0001-dependency-injection.md:578]: @bjhham so no support for bean override + parallel testing?
Or it's a new context for each test?

Comment (bjhham) [reply proposals/0001-dependency-injection.md:578]: Yes, it's a new context for every test so there's support, but heavy instantiation will cause a performance hit.  I'll make sure to log a ticket for this.

File: proposals/0001-dependency-injection.md [hunk 15]
```diff
@@ -0,0 +1,642 @@
...
+Here are the examples of how to resolve dependencies from a Ktor module:
+
+```kotlin
+fun Application.configureRouting() {
+    // Resolve by property delegation
+    val users: Repository<User> by dependencies
+    // Using a named instance
+    val mongo: DataSource by dependencies.named("mongo")
+    // Resolve from the dependencies property in the Application scope
+    val messages: Repository<Message> = dependencies.resolve()
```
Comment (osipxd) [inline proposals/0001-dependency-injection.md]: Should we mention here that dependencies are instantiated lazily when delegate is used and eagerly when `resolve()` is used?

File: proposals/0001-dependency-injection.md (hunk 13 above)
Comment (osipxd) [reply proposals/0001-dependency-injection.md:479]: @arnaudgiuliani, here is a link to track this issue:
[KT-73657](https://youtrack.jetbrains.com/issue/KT-73657) "Provide a way to choose reflection implementation explicitly: kotlin-reflect or stdlib"

**Review by osipxd — APPROVED** [review]

File: proposals/0001-dependency-injection.md (hunk 14 above)
Comment (marychatte) [reply proposals/0001-dependency-injection.md:392]: Can we ensure that users don’t unintentionally rely on implicit injection during resolution? Will there be any explicit checks with warnings or documentation to encourage the use of `create()` instead of `resolve()` for automatic construction?

File: proposals/0001-dependency-injection.md [hunk 16]
```diff
@@ -0,0 +1,642 @@
...
+The standard behavior for `resolve()` will be to check the instance repository for the provided key and fail if nothing 
+is found.  For some projects, developers may want to override this behavior by creating new instances, or searching 
+some other source, instead.
+
+## Validation
+[validation]: #validation
+
+Normally, validation occurs when calling `resolve()` in the application.  This will make an attempt to find the instance
+of the expected type in the instance repository.  If nothing was declared for this type, then it will throw an exception
+`DependencyMissingException`, unless otherwise implemented by a custom provider.  During declaration, there is also a
```
Comment (marychatte) [inline proposals/0001-dependency-injection.md:421]: How will we handle parameter resolution when certain parameters are optional or have default values? Will there be a configuration for how to deal with such cases?

File: proposals/0001-dependency-injection.md [hunk 17]
```diff
@@ -0,0 +1,642 @@
...
+application fails due to missing dependencies.  For this, some DI frameworks have introduced compile-time checking to 
+ensure that all dependencies can be resolved before you run the application.
+
+### Compile-time validation
+[compile-time-validation]: #compile-time-validation
+
+For Kotlin applications, compile-time validation of sources will usually involve gradle extensions.  Since Ktor already
+has a Gradle plugin, this seems like a natural fit for handling this kind of validation.  We should be able to check 
+the classpath for the dependency injection package, then run the validation accordingly after the compilation step.  The 
+validation can come in the form of a compiler plugin.
```
Comment (marychatte) [inline proposals/0001-dependency-injection.md:435]: Will we add built-in logging to visualize the dependency resolution process?

**Review by marychatte — APPROVED** [review]
Great work!

File: proposals/0001-dependency-injection.md (hunk 13 above)
Comment (arnaudgiuliani) [reply proposals/0001-dependency-injection.md:479]: thanks @osipxd

File: proposals/0001-dependency-injection.md (hunk 16 above)
Comment (bjhham) [reply proposals/0001-dependency-injection.md:421]: Good point!  I agree, we'll need some configurable logic to handle both these conditions.  I think the default behaviour would be to try supplying the dependency and using the default when it's missing.

File: proposals/0001-dependency-injection.md (hunk 17 above)
Comment (bjhham) [reply proposals/0001-dependency-injection.md:435]: Yeah, I'd say we ought to provide something there in the debug logs to clearly show things are wired up.  The Koin IDEA plugin includes a panel where you can click on different items to see where it's declared in the code - would be nice to include something similar in the Ktor plugin, though it'd be pretty fancy.

File: proposals/0001-dependency-injection.md (hunk 14 above)
Comment (bjhham) [reply proposals/0001-dependency-injection.md:392]: So my thinking is that `resolve()` will only draw from the declared instances and `create()` will be the only means for issuing the automatic construction.

**Review by e5l — APPROVED** [review]

**Comment by e5l** [comment]
Sorry for the delay, lgtm!
